truffle/src/main/java/org/jruby/truffle/translator/BodyTranslator.java

```diff
@@ -495,21 +522,46 @@ private RubyNode translateRubiniusInvokePrimitive(SourceSection sourceSection, C
 
         final RubiniusPrimitiveConstructor primitive = context.getRubiniusPrimitiveManager().getPrimitive(primitiveName);
 
-        final List<RubyNode> arguments = new ArrayList<>();
+        if (primitive instanceof RubiniusPrimitiveNodeConstructor) {
+            final RubiniusPrimitiveNodeConstructor nodeConstructor = (RubiniusPrimitiveNodeConstructor) primitive;
+
+            final List<RubyNode> arguments = new ArrayList<>();
 
-        // The first argument was the symbol so we ignore it
-        for (int n = 1; n < node.getArgsNode().childNodes().size(); n++) {
-            RubyNode readArgumentNode = node.getArgsNode().childNodes().get(n).accept(this);
+            // The first argument was the symbol so we ignore it
+            for (int n = 1; n < node.getArgsNode().childNodes().size(); n++) {
+                RubyNode readArgumentNode = node.getArgsNode().childNodes().get(n).accept(this);
 
-            if (ArrayUtils.contains(primitive.getAnnotation().lowerFixnumParameters(), n)) {
-                readArgumentNode = new FixnumLowerNode(readArgumentNode);
+                if (ArrayUtils.contains(nodeConstructor.getAnnotation().lowerFixnumParameters(), n)) {
+                    readArgumentNode = new FixnumLowerNode(readArgumentNode);
+                }
+
+                arguments.add(readArgumentNode);
             }
 
-            arguments.add(readArgumentNode);
+            return new InvokeRubiniusPrimitiveNode(context, sourceSection,
+                    nodeConstructor.getFactory().createNode(context, sourceSection, arguments.toArray(new RubyNode[arguments.size()])));
+        } else if (primitive instanceof RubiniusPrimitiveCallConstructor) {
+            final RubiniusPrimitiveCallConstructor callConstructor = (RubiniusPrimitiveCallConstructor) primitive;
+
+            final List<RubyNode> arguments = new ArrayList<>();
+
+            // The first argument was the symbol so we ignore it
+            for (int n = 1; n < node.getArgsNode().childNodes().size(); n++) {
+                RubyNode readArgumentNode = node.getArgsNode().childNodes().get(n).accept(this);
+                arguments.add(readArgumentNode);
+            }
+
+            return new RubyCallNode(
+                    context,
+                    sourceSection,
+                    callConstructor.getMethod(),
+                    new LiteralNode(context, sourceSection, callConstructor.getModule()),
+                    null,
+                    false,
+                    arguments.toArray(new RubyNode[arguments.size()]));
```

Comment:
It's only the arguments array that can be shared. I didn't do it originally as the body of the loop is a bit different, but I can probably do a second pass to handle that.